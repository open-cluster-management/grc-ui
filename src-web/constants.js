/* Copyright (c) 2020 Red Hat, Inc. */

export default {
  EMPTY_CHOICE: 'EMPTY_CHOICE',
  overviewTopInfoClustersStr: 'string.clusters',
  overviewTopInfoAppStr: 'string.applications'
}
