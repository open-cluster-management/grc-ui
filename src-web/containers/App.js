/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import resources from '../../lib/shared/resources'
resources(() => {
  require('../../scss/common.scss')
})
import PropTypes from 'prop-types'
// without curly braces means component with redux
// eslint-disable-next-line import/no-named-as-default
import SecondaryHeader from '../components/modules/SecondaryHeader'
import { Route, Switch, Redirect, withRouter } from 'react-router-dom'
import client from '../../lib/shared/client'
import config from '../../lib/shared/config'
import Modal from '../components/common/Modal'
import GrcRouter from './GrcRouter'
import loadable from '@loadable/component'
import { LocaleContext } from '../components/common/LocaleContext'
import { AcmHeader, AcmRoute } from '@open-cluster-management/ui-components'
import WelcomeStatic from './Welcome'
import GrcApolloClient from '../../lib/client/apollo-client'
import { userAccessSuccess } from '../actions/access'
import { connect } from 'react-redux'

export const ResourceToolbar = loadable(() => import(/* webpackChunkName: "ResourceToolbar" */ '../components/common/ResourceToolbar'))

class App extends React.Component {

  static propTypes = {
    match: PropTypes.object,
    staticContext: PropTypes.object,
    url: PropTypes.string,
  }

  constructor(props) {
    super(props)
    if (client && document.getElementById('propshcm')) {
      this.serverProps = JSON.parse(document.getElementById('propshcm').textContent)
    }
  }

  getServerProps() {
    if (client) {
      return this.serverProps
    }
    return this.props.staticContext
  }

  componentDidMount() {
    GrcApolloClient.getUserAccess().then((response) => {
      this.props.updateUserAccess(response.data.items)
    })
  }

  render() {
    const serverProps = this.getServerProps()
    const { match } = this.props
    return (
      <LocaleContext.Provider value={serverProps.context}>
        <div className='expand-vertically'>
          <SecondaryHeader />
          <ResourceToolbar />
          <Switch>
            <Route path={`${match.url}`} render={() => <GrcRouter />} />
            <Redirect to={`${config.contextPath}`} />
          </Switch>
          <Modal locale={serverProps.context.locale} />
        </div>
      </LocaleContext.Provider>
    )
  }
}

App.propTypes = {
  updateUserAccess: PropTypes.func
}

const mapDispatchToProps = (dispatch) => {
  return {
    updateUserAccess: (access) => dispatch(userAccessSuccess(access))
  }
}

const AppWithUserAccess = withRouter(connect(null, mapDispatchToProps)(App))

const getAcmRoute = (props) => {
  let path = ''
  if (client) {
    path = window.location.pathname
  } else {
    path = props.url
  }
  if (path.includes(config.contextPath)) {
    return AcmRoute.GovernRisk
  }
  return AcmRoute.Welcome
}

// eslint-disable-next-line react/display-name
export default props => (
  // eslint-disable-next-line react/prop-types
  <AcmHeader route={getAcmRoute(props)} >
    <Route path={config.contextPath} serverProps={props} component={AppWithUserAccess} />
    <Route path={'/multicloud/welcome'} serverProps={props} component={WelcomeStatic} />
  </AcmHeader>
)
