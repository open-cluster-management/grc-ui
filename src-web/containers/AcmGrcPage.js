/* eslint-disable react/prop-types */
/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import { useQuery } from '@apollo/client'
import { AcmPageContent, AcmPageHeader, AcmAutoRefreshSelect, AcmRefreshTime, AcmSecondaryNav, AcmAlert } from '@open-cluster-management/ui-components'
import { Spinner } from '@patternfly/react-core'
import { useHistory, useParams } from 'react-router-dom'

import { INITIAL_POLL_INTERVAL, REFRESH_INTERVALS, REFRESH_INTERVAL_COOKIE } from '../../lib/shared/constants'
import { getPageDefinition } from './definitions/PageDefinitions'

let timestamp = new Date().toString()

function AcmGrcPage(props) {
  const allProps = {...props, ...useParams(), history: useHistory()}
  const page = getPageDefinition(allProps)
  const { loading, data={}, refetch, error, previousData } = useQuery(page.query, { variables: page.query_variables, notifyOnNetworkStatusChange: true })
  const { items } = data
  if (!loading) {
    timestamp = new Date().toString()
  }
  return (
    <React.Fragment>
      <AcmPageHeader title={page.title}
        breadcrumb={page.breadcrumb}
        navigation={page.navigation && (
          <AcmSecondaryNav>
            {page.navigation.map(nav => nav(allProps))}
          </AcmSecondaryNav>
        )}
        controls={
          <React.Fragment>
            {page.refreshControls && (
              <React.Fragment>
                <AcmAutoRefreshSelect refetch={refetch}
                  refreshIntervals={REFRESH_INTERVALS}
                  refreshIntervalCookie={REFRESH_INTERVAL_COOKIE}
                  initPollInterval={INITIAL_POLL_INTERVAL} />
                <AcmRefreshTime timestamp={timestamp} reloading={loading} />
              </React.Fragment>
            )}
            <div className='page-header-button-group'>
            {page.buttons && (
              page.buttons.map(btn => btn(allProps))
            )}
            </div>
          </React.Fragment>}>
      </AcmPageHeader>
      <AcmPageContent id={page.id}>
        {(() => {
          if(error) {
            return <AcmAlert isInline={true} variant='danger'
              subtitle={error} />
          }
          if (loading && !previousData || items === undefined ) {
            return <Spinner className='patternfly-spinner' />
          } else {
            return page.children({ items })
          }
        })()}
      </AcmPageContent>
    </React.Fragment>
  )
}

export default AcmGrcPage
