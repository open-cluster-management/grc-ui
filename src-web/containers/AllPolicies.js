/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Query } from 'react-apollo'
import { withRouter } from 'react-router-dom'
import { AcmButton, AcmPage, AcmPageHeader, AcmAutoRefreshSelect, AcmRefreshTime } from '@open-cluster-management/ui-components'
import { ALL_POLICIES } from '../../lib/client/queries'
import { LocaleContext } from '../components/common/LocaleContext'
import { INITIAL_REFRESH_TIME, REFRESH_INTERVALS, REFRESH_INTERVAL_COOKIE } from '../../lib/shared/constants'
import msgs from '../../nls/platform.properties'
import config from '../../lib/shared/config'
import Page from '../components/common/Page'
// eslint-disable-next-line import/no-named-as-default
import GrcView from '../components/modules/GrcView'
import checkCreatePermission from '../components/common/CheckCreatePermission'


class AllPolicies extends React.Component {

  static contextType = LocaleContext

  constructor (props) {
    super(props)
    this.firstLoad = true
  }

  render () {
    const { locale } = this.context
    const { history, userAccess } = this.props
    const pollInterval = localStorage.getItem(REFRESH_INTERVAL_COOKIE) || INITIAL_REFRESH_TIME
    return (
      <Page>
        <Query query={ALL_POLICIES} pollInterval={pollInterval*1000} notifyOnNetworkStatusChange >
          {( complianceResult ) => {
            const {data={}, loading, refetch} = complianceResult
            const { items } = data
            const error = items ? null : complianceResult.error
            if (error) {
              const errorName = complianceResult.error.graphQLErrors[0].name ? complianceResult.error.graphQLErrors[0].name : error.name
              error.name = errorName
            }
            if (!loading) {
              this.timestamp = new Date().toString()
            }
            const isDisabled = checkCreatePermission(userAccess)===0
            return (
              <AcmPage>
                <AcmPageHeader title= {msgs.get('routes.grc', locale)} controls={
                  <React.Fragment>
                    <AcmAutoRefreshSelect refetch={refetch}
                      refreshIntervals={REFRESH_INTERVALS}
                      refreshIntervalCookie={REFRESH_INTERVAL_COOKIE}
                      initRefreshTime={INITIAL_REFRESH_TIME} />
                    <AcmRefreshTime timestamp={this.timestamp} reloading={loading} />
                    <AcmButton id='create-policy' isDisabled={isDisabled}
                      tooltip={msgs.get('error.permission.disabled', locale)}
                      onClick={() => history.push(`${config.contextPath}/create`)}>
                      {msgs.get('routes.create.policy', locale)}
                    </AcmButton>
                  </React.Fragment>
                }>
                </AcmPageHeader>
                <GrcView
                  loading={!items && loading}
                  error={error}
                  grcItems={items}
                />
              </AcmPage>
            )
          }}
        </Query>
      </Page>
    )
  }
}
AllPolicies.propTypes = {
  history: PropTypes.object.isRequired,
  userAccess: PropTypes.array,
}


const mapStateToProps = (state) => {
  const userAccess = state.userAccess && state.userAccess.access
    ? state.userAccess.access
    : []
  return {
    userAccess: userAccess,
  }
}

export default withRouter(connect(mapStateToProps)(AllPolicies))
