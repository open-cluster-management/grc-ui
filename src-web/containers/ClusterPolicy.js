/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************
 * Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateSecondaryHeader} from '../actions/common'
import {getPollInterval} from '../components/common/RefreshTimeSelect'
import lodash from 'lodash'
import { GRC_REFRESH_INTERVAL_COOKIE, RESOURCE_TYPES } from '../../lib/shared/constants'
import msgs from '../../nls/platform.properties'
import { Query } from 'react-apollo'
import { HCMPolicy } from '../../lib/client/queries'
import getResourceDefinitions from '../definitions'
import PolicyClusterDetail from '../components/common/PolicyClusterDetail'

class ClusterPolicy extends React.Component {
  static propTypes = {
    location: PropTypes.object,
    params: PropTypes.object,
    resourceType: PropTypes.object,
    updateSecondaryHeader: PropTypes.func,
  }

  static defaultProps = {
    resourceType: RESOURCE_TYPES.HCM_COMPLIANCES,
  }

  static contextTypes = {
    locale: PropTypes.string
  }

  constructor (props) {
    super(props)
  }

  componentWillMount() {
    const { updateSecondaryHeader:localUpdateSecondaryHeader } = this.props
    localUpdateSecondaryHeader(this.getPolicyName(true), null, this.getBreadcrumb())
  }

  getPolicyName(withParentNamespace) {
    const { location } = this.props,
          urlSegments = location.pathname.split('/')
    if(withParentNamespace){
      return urlSegments[urlSegments.length - 1]
    }
    else{
      const nameSegments = urlSegments[urlSegments.length - 1].split('.')
      return nameSegments[1]
    }
  }

  getRootPolicyNamespace() {
    const { location } = this.props,
          urlSegments = location.pathname.split('/'),
          nameSegments = urlSegments[urlSegments.length - 1].split('.')
    return nameSegments[0]
  }

  getClusterName() {
    const { location } = this.props,
          urlSegments = location.pathname.split('/')
    return urlSegments[urlSegments.length - 2]
  }

  getBreadcrumb() {
    const breadcrumbItems = []
    const {  location, resourceType } = this.props,
          { locale } = this.context,
          urlSegments = location.pathname.split('/')

    // Push only one breadcrumb to overview page
    if (resourceType.name === RESOURCE_TYPES.HCM_COMPLIANCES.name) {
      breadcrumbItems.push({
        label: msgs.get('tabs.hcmcompliance', locale),
        noLocale: true,
        url: `${urlSegments.slice(0, 3).join('/')}/all`
      },
      {
        label: this.getPolicyName(false),
        noLocale: true,
        url: `${urlSegments.slice(0, 3).join('/')}/all/${this.getRootPolicyNamespace()}/${this.getPolicyName(false)}`
      },
      {
        label: this.getClusterName(),
        noLocale: true,
        url: `${urlSegments.join('/')}`
      }
      )
    }
    return breadcrumbItems
  }

  render() {
    const url = lodash.get(this.props, 'location.pathname')
    const urlSegments = url.split('/')
    const policyName = urlSegments[urlSegments.length - 1]
    const policyNamespace = urlSegments[urlSegments.length - 2]
    const pollInterval = getPollInterval(GRC_REFRESH_INTERVAL_COOKIE)
    const {params, resourceType} = this.props
    const staticResourceData = getResourceDefinitions(resourceType)
    return (
      <Query query={HCMPolicy} variables={{name: policyName, clusterName: policyNamespace}} pollInterval={pollInterval} notifyOnNetworkStatusChange >
        {(result) => {
          const {data={}, loading, startPolling, stopPolling, refetch} = result
          const { policies } = data
          const error = policies ? null : result.error
          const firstLoad = this.firstLoad
          this.firstLoad = false
          const reloading = !firstLoad && loading
          if (!reloading) {
            this.timestamp = new Date().toString()
          }
          const refreshControl = {
            reloading,
            refreshCookie: GRC_REFRESH_INTERVAL_COOKIE,
            startPolling, stopPolling, refetch,
            timestamp: this.timestamp
          }

          return (
            <PolicyClusterDetail
              resourceType={resourceType}
              staticResourceData={staticResourceData}
              policies={policies}
              loading={!policies && loading}
              error={error}
              refreshControl={refreshControl}
              params={params}
            />)
        }}
      </Query>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    updateSecondaryHeader: (title, tabs, breadcrumbItems, links) => dispatch(updateSecondaryHeader(title, tabs, breadcrumbItems, links))
  }
}

export default withRouter(connect(null, mapDispatchToProps)(ClusterPolicy))
