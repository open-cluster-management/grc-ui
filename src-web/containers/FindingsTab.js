/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { Query } from 'react-apollo'
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import {GRC_REFRESH_INTERVAL_COOKIE} from '../../lib/shared/constants'
import {getPollInterval} from '../components/common/RefreshTimeSelect'
import Page from '../components/common/Page'
// without curly braces means component with redux
// eslint-disable-next-line import/no-named-as-default
import GrcView from '../components/GrcView'
import {updateSecondaryHeader} from '../actions/common'
import { FindingList } from '../../lib/client/queries'
import msgs from '../../nls/platform.properties'
import _ from 'lodash'

export class FindingsTab extends React.Component {

  static propTypes = {
    secondaryHeaderProps: PropTypes.object,
    updateSecondaryHeader: PropTypes.func,
  }

  constructor (props) {
    super(props)
    this.firstLoad = true
  }

  componentDidMount() {
    const { updateSecondaryHeader:localUpdateSecondaryHeader, secondaryHeaderProps } = this.props
    const { title, tabs, links, information } = secondaryHeaderProps
    localUpdateSecondaryHeader(msgs.get(title, this.context.locale), tabs, links, msgs.get(information, this.context.locale))
  }

  render () {
    const { secondaryHeaderProps, userpreferences } = this.props
    const activeAccountId = (userpreferences && userpreferences.userPreferences) ? userpreferences.userPreferences.activeAccountId : ''
    const pollInterval = getPollInterval(GRC_REFRESH_INTERVAL_COOKIE)
    return (
      <Page>
        <Query query={FindingList} variables={{userAccountID: activeAccountId}} pollInterval={pollInterval} notifyOnNetworkStatusChange >
          {( result ) => {
            const {data={}, loading, startPolling, stopPolling, refetch} = result
            const { findings } = data
            const error = findings ? null : result.error
            const firstLoad = this.firstLoad
            this.firstLoad = false
            const reloading = !firstLoad && loading
            if (!reloading) {
              this.timestamp = new Date().toString()
            }
            const refreshControl = {
              reloading,
              refreshCookie: GRC_REFRESH_INTERVAL_COOKIE,
              startPolling, stopPolling, refetch,
              timestamp: this.timestamp
            }

            return (
              <div>
                <GrcView
                  loading={!findings && loading}
                  error={error}
                  grcItems={this.formatEmptySecurityClassification(findings)}
                  refreshControl={refreshControl}
                  secondaryHeaderProps={secondaryHeaderProps}
                />
              </div>
            )
          }
          }
        </Query>
      </Page>
    )
  }

  formatEmptySecurityClassification(findings) {
    if(findings){
      findings.forEach((finding) => {
        if(!_.get(finding, 'securityClassification', null)) {
          finding['securityClassification'] = {
            'securityStandards': [''],
            'securityCategories': [''],
            'securityControl': ''
          }
          return finding['securityClassification']
        }
        return undefined
      })
    }
    return findings
  }
}

FindingsTab.propTypes = {
  userpreferences: PropTypes.object,
}

const mapStateToProps = state => {
  return {
    userpreferences: state.userpreferences
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    updateSecondaryHeader: (title, tabs, links, information) => dispatch(updateSecondaryHeader(title, tabs, undefined, links, '', information)),
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(FindingsTab))
