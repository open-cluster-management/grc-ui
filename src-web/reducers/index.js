/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/*
 *
 * Things you should never do inside a reducer:
 *
 * - Mutate its arguments
 * - Perform side effects like API calls and routing transitions
 * - Call non-pure functions, e.g. Date.now() or Math.random()
 *
 * Reducers must be deterministic pure functions.  Given the same arguments, it should calculate the next state and return it.
 * No surprises. No side effects. No API calls. No mutations. Just a calculation.
 *
 * Selectors should sit along side reducers.
 */

import { createResourceReducer, resourceReducerFunction } from './common'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import lodash from 'lodash'

export { user, loggedIn } from './user'
export { uiconfig } from './uiconfig'
export { role } from './role'
export { userpreferences } from './userpreferences'
export { resourceToolbar } from './common'
export { secondaryHeader } from './common'

export { modal } from './modal'

export const HCMApplicationList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_APPLICATIONS))
export const HCMComplianceList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_COMPLIANCES))
export const HCMPolicyList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_POLICIES))
export const HCMPolicyApplicationList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_POLICIES_PER_APPLICATION))
export const HCMPolicyPolicyList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_POLICIES_PER_POLICY))
export const HCMPolicyClusterList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_POLICIES_PER_CLUSTER))
export const HCMSecurityFindingsList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_SECURITY_FINDINGS))
export const HCMClusterFindingsList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_CLUSTER_FINDINGS))

// export { default as catalog } from './catalog'
export { resourceFilters } from './filter'

export function predicate(resourceType, action) {
  if (lodash.isEqual(resourceType, action.resourceType)) {
    return true
  }
  const result = lodash.find(lodash.values(resourceType), type => {
    if (typeof type === 'string') {
      return type.indexOf(action.resourceType) > -1
    }
    return false
  })

  return result ? result : null
}
