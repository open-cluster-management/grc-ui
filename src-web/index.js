/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
import React from 'react'
import { render } from 'react-dom'
import { Provider } from 'react-redux'
import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import { BrowserRouter } from 'react-router-dom'
import { ApolloProvider } from 'react-apollo'
import { App } from './containers/App'
import * as reducers from './reducers'
import config from '../lib/shared/config'
import GrcApolloClient from '../lib/client/apollo-client'
import ScrollToTop from './components/common/ScrollToTop'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

const loggerMiddleware = createLogger()
// Grab the state from a global variable injected into the server-generated HTML
const preloadedState = window.__PRELOADED_STATE__

const middleware = [thunkMiddleware] // lets us dispatch() functions
if (!window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ && config['featureFlags:reduxLogger']) {
  middleware.push(loggerMiddleware) // middleware that logs actions
}

// Allow the passed state to be garbage-collected
delete window.__PRELOADED_STATE__

// Create Redux store with initial state
const store = createStore(combineReducers(reducers), preloadedState, composeEnhancers(
  applyMiddleware(...middleware)
))



render(
  <ApolloProvider client={GrcApolloClient.getGrcClient()}>
    <Provider store={store}>
      <BrowserRouter>
        <ScrollToTop>
          <App />
        </ScrollToTop>
      </BrowserRouter>
    </Provider>
  </ApolloProvider>
  , document.getElementById('page')
)
