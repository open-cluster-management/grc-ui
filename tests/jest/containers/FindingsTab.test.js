/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import React from 'react'
import {FindingsTab} from '../../../src-web/containers/FindingsTab'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import * as reducers from '../../../src-web/reducers'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import thunkMiddleware from 'redux-thunk'
import GrcApolloClient from '../../../lib/client/apollo-client'
import { ApolloProvider } from 'react-apollo'
import { Provider } from 'react-redux'
import { shallow } from 'enzyme'
import { findingsData } from './ContainersTestingData'

describe('FindingsTab container test', () => {
  it('renders as expected', () => {
    const preloadedState = window.__PRELOADED_STATE__
    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), preloadedState, composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const secondaryHeaderProps = {
      'title': 'routes.create.policy',
      'information': 'policy.create.tooltip',
      'breadcrumbItems': [
        {
          'id': 'policy-overview',
          'label': 'routes.grc',
          'url': '/multicloud/policies'
        },
        {
          'id': 'policy-overview-all',
          'label': 'routes.policies',
          'url': '/multicloud/policies/all'
        }
      ]
    }
    const updateSecondaryHeader= jest.fn()
    const component = renderer.create(
      <ApolloProvider client={GrcApolloClient.getGrcClient()}>
        <Provider store={store}>
          <BrowserRouter>
            <FindingsTab
              secondaryHeaderProps={secondaryHeaderProps}
              updateSecondaryHeader={updateSecondaryHeader}
            />
          </BrowserRouter>
        </Provider>
      </ApolloProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})


describe('FindingsTab container formatEmptySecurityClassification test', () => {
  it('renders as expected', () => {
    const secondaryHeaderProps = {
      'title': 'routes.create.policy',
      'information': 'policy.create.tooltip',
      'breadcrumbItems': [
        {
          'id': 'policy-overview',
          'label': 'routes.grc',
          'url': '/multicloud/policies'
        },
        {
          'id': 'policy-overview-all',
          'label': 'routes.policies',
          'url': '/multicloud/policies/all'
        }
      ]
    }
    const updateSecondaryHeader= jest.fn()
    const component = shallow(
      <FindingsTab
        secondaryHeaderProps={secondaryHeaderProps}
        updateSecondaryHeader={updateSecondaryHeader}
      />
    )
    expect(component.instance().formatEmptySecurityClassification(null)).toMatchSnapshot()

    expect(component.instance().formatEmptySecurityClassification(findingsData)).toMatchSnapshot()
  })
})

