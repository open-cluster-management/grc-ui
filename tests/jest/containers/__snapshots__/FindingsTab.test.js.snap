// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindingsTab container formatEmptySecurityClassification test renders as expected 1`] = `null`;

exports[`FindingsTab container formatEmptySecurityClassification test renders as expected 2`] = `
Array [
  Object {
    "__typename": "Occurrence",
    "context": Object {
      "accountId": "id-mycluster-account",
      "clusterName": "clusterhub",
      "namespaceName": "Excludes: [kube-*], Includes: [default]",
      "region": "clusterhub",
      "resourceCrn": null,
      "resourceId": "b35230ab-fa76-11e9-bf6c-005056a0b88e",
      "resourceName": "policy-iampolicy",
      "resourceType": "Policy",
      "serviceCrn": null,
      "serviceName": "icp4mcm-findings",
    },
    "finding": Object {
      "certainty": "HIGH",
      "dataTransferred": null,
      "networkConnection": null,
      "nextSteps": Array [
        Object {
          "title": "View the details for the policy violation using the Security findings dashboard.",
          "url": null,
        },
      ],
      "severity": "MEDIUM",
    },
    "longDescription": "The Multicloud Manager Policy violation caused a security finding to be created",
    "name": "id-mycluster-account/providers/security-advisor/occurrences/clusterhub-policy-b35230ab-fa76-11e9-bf6c-005056a0b88e",
    "providerId": "security-advisor",
    "providerName": "id-mycluster-account/providers/security-advisor",
    "remediation": "NonCompliant; 2 clusterrole admin users violations detected in namespace \`cluster-wide\` ; 0 rolebindings violations detected in namespace \`default\`, violated rolebinding : []; 0 rolebindings violations detected in namespace \`kube-public\`, violated rolebinding : []",
    "reportedBy": Object {
      "id": "mcm-policy-adapter",
      "title": "Security Advisor Multicloud Manager Policy Findings Adapter",
      "url": null,
    },
    "securityClassification": Object {
      "securityCategories": Array [
        "",
      ],
      "securityControl": "",
      "securityStandards": Array [
        "",
      ],
    },
    "shortDescription": "Policy violation finding",
    "updateTime": "2019-10-30T17:19:02.276706Z",
  },
  Object {
    "__typename": "Occurrence",
    "context": Object {
      "accountId": "id-mycluster-account",
      "clusterName": "cluster1",
      "namespaceName": "Excludes: [kube-*], Includes: [default]",
      "region": "cluster1",
      "resourceCrn": null,
      "resourceId": "b35230ab-fa76-11e9-bf6c-005056a0b88e",
      "resourceName": "policy-iampolicy",
      "resourceType": "Policy",
      "serviceCrn": null,
      "serviceName": "icp4mcm-findings",
    },
    "finding": Object {
      "certainty": "HIGH",
      "dataTransferred": null,
      "networkConnection": null,
      "nextSteps": Array [
        Object {
          "title": "View the details for the policy violation using the Security findings dashboard.",
          "url": null,
        },
      ],
      "severity": "MEDIUM",
    },
    "longDescription": "The Multicloud Manager Policy violation caused a security finding to be created",
    "name": "id-mycluster-account/providers/security-advisor/occurrences/cluster1-policy-b35230ab-fa76-11e9-bf6c-005056a0b88e",
    "providerId": "security-advisor",
    "providerName": "id-mycluster-account/providers/security-advisor",
    "remediation": "NonCompliant; 1 clusterrole admin users violations detected in namespace \`cluster-wide\` ; 0 rolebindings violations detected in namespace \`default\`, violated rolebinding : []; 0 rolebindings violations detected in namespace \`kube-public\`, violated rolebinding : []",
    "reportedBy": Object {
      "id": "mcm-policy-adapter",
      "title": "Security Advisor Multicloud Manager Policy Findings Adapter",
      "url": null,
    },
    "securityClassification": Object {
      "securityCategories": Array [
        "IdentityManagementAndAccessControl",
      ],
      "securityControl": "Iam",
      "securityStandards": Array [
        "NIST-CSF",
      ],
    },
    "shortDescription": "Policy violation finding",
    "updateTime": "2019-10-30T17:18:57.777080Z",
  },
  Object {
    "__typename": "Occurrence",
    "context": Object {
      "accountId": "id-mycluster-account",
      "clusterName": "clusterhub",
      "namespaceName": "Excludes: [kube-*], Includes: [default]",
      "region": "clusterhub",
      "resourceCrn": null,
      "resourceId": "0efa8d02-f9ee-11e9-bf6c-005056a0b88e",
      "resourceName": "policy-pod",
      "resourceType": "Policy",
      "serviceCrn": null,
      "serviceName": "icp4mcm-findings",
    },
    "finding": Object {
      "certainty": "HIGH",
      "dataTransferred": null,
      "networkConnection": null,
      "nextSteps": Array [
        Object {
          "title": "View the details for the policy violation using the Security findings dashboard.",
          "url": null,
        },
      ],
      "severity": "HIGH",
    },
    "longDescription": "The Multicloud Manager Policy violation caused a security finding to be created",
    "name": "id-mycluster-account/providers/security-advisor/occurrences/clusterhub-policy-0efa8d02-f9ee-11e9-bf6c-005056a0b88e",
    "providerId": "security-advisor",
    "providerName": "id-mycluster-account/providers/security-advisor",
    "remediation": "pods \`nginx-pod\` is missing",
    "reportedBy": Object {
      "id": "mcm-policy-adapter",
      "title": "Security Advisor Multicloud Manager Policy Findings Adapter",
      "url": null,
    },
    "securityClassification": Object {
      "securityCategories": Array [
        "SecurityContinuousMonitoring",
      ],
      "securityControl": "Other",
      "securityStandards": Array [
        "NIST-CSF",
      ],
    },
    "shortDescription": "Policy violation finding",
    "updateTime": "2019-10-30T17:18:54.164334Z",
  },
  Object {
    "__typename": "Occurrence",
    "context": Object {
      "accountId": "id-mycluster-account",
      "clusterName": "cluster1",
      "namespaceName": "Excludes: [kube-*], Includes: [default]",
      "region": "cluster1",
      "resourceCrn": null,
      "resourceId": "0efa8d02-f9ee-11e9-bf6c-005056a0b88e",
      "resourceName": "policy-pod",
      "resourceType": "Policy",
      "serviceCrn": null,
      "serviceName": "icp4mcm-findings",
    },
    "finding": Object {
      "certainty": "HIGH",
      "dataTransferred": null,
      "networkConnection": null,
      "nextSteps": Array [
        Object {
          "title": "View the details for the policy violation using the Security findings dashboard.",
          "url": null,
        },
      ],
      "severity": "HIGH",
    },
    "longDescription": "The Multicloud Manager Policy violation caused a security finding to be created",
    "name": "id-mycluster-account/providers/security-advisor/occurrences/cluster1-policy-0efa8d02-f9ee-11e9-bf6c-005056a0b88e",
    "providerId": "security-advisor",
    "providerName": "id-mycluster-account/providers/security-advisor",
    "remediation": "pods \`nginx-pod\` is missing",
    "reportedBy": Object {
      "id": "mcm-policy-adapter",
      "title": "Security Advisor Multicloud Manager Policy Findings Adapter",
      "url": null,
    },
    "securityClassification": Object {
      "securityCategories": Array [
        "",
      ],
      "securityControl": "",
      "securityStandards": Array [
        "",
      ],
    },
    "shortDescription": "Policy violation finding",
    "updateTime": "2019-10-30T17:18:50.849184Z",
  },
]
`;

exports[`FindingsTab container test renders as expected 1`] = `
<div
  className="page-content-container"
  role="main"
>
  <div>
    <div
      aria-live="assertive"
      className="bx--loading content-spinner"
    >
      <svg
        className="bx--loading__svg"
        viewBox="-75 -75 150 150"
      >
        <title>
          Active loading indicator
        </title>
        <circle
          className={null}
          cx="0"
          cy="0"
          r="37.5"
        />
      </svg>
    </div>
  </div>
</div>
`;
